content/6_Advanced/DP_SOS.mdx

```diff
@@ -26,6 +26,85 @@ frequency: 0
 	</Resource>
 </Resources>
 
+
+Sum over subsets (SOS) dp is a trick that allows you to efficiently compute the sum of all the subsets of an array.
+
+The naive solution would be to iterate through every other mask and if the other mask is a subset of the current mask we add it to the current answer.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
+vector<int> a(1 << n);
+vector<int> sos(1 << n);
+for (int i = 0; i < (1 << n); i++) {
+	// iterate over all subsets of i
+	for (int j = 0; j < (1 << n); j++) {
+		if ((i & j) == j) { sos[i] += a[j]; }
+	}
+}
+```
+</CPPSection>
+</LanguageSection>
+
+
+We can speed this up if we iterate over only subsets of the current mask and add up all of the those values to get the sum over subsets for a particular mask.
+The time complexity here is O($3^n$)
+
+<LanguageSection>
+<CPPSection>
+```cpp
+int n = 20;
+vector<int> a(1 << n);
+vector<int> sos(1 << n);
+for (int i = 0; i < (1 << n); i++) {
+	// iterate over all subsets of i
+	for (int j = (i - 1) & i; j >= 0; j = (j - 1) & i) { sos[i] += a[j]; }
+}
+```
+</CPPSection>
+</LanguageSection>
+
+
+Notice how in both of these examples we don't seem to be saving much information between different queries which is the essence of DP.
+Define $SOS(mask, x)$ to be the sum of subsets of $mask$ such that the first $x$ bits of the subset are identical to the first $x$ bits of mask.
```

Comment:
```suggestion
Define $\texttt{SOS}(mask, x)$ to be the sum of subsets of $mask$ such that the first $x$ bits of the subset are identical to the first $x$ bits of mask.
```
texttt-ify the rest as aswell

generally variable names $\geq 3$ chars should be texttt-ified